File: Pharmaceutical_Company_Example/Behavioral/R/Chain_of_Responsibility.R

# Base class for a drug approval request handler
DrugApprovalHandler <- setRefClass(
  "DrugApprovalHandler",
  fields = list(next_handler = "ANY"),
  methods = list(
    set_next = function(handler) {
      .self$next_handler <- handler
      return(handler)
    },
    handle_request = function(drug) {
      stop("This method should be implemented by subclasses.")
    },
    pass_to_next = function(drug) {
      if (!is.null(.self$next_handler)) {
        .self$next_handler$handle_request(drug)
      } else {
        message("End of chain, no handler could process the request.")
      }
    }
  )
)

# Concrete handler for Initial Review
InitialReviewHandler <- setRefClass(
  "InitialReviewHandler",
  contains = "DrugApprovalHandler",
  methods = list(
    handle_request = function(drug) {
      if (drug$initial_review_status == "Pending") {
        message("Initial review passed for ", drug$name)
        drug$initial_review_status <- "Approved"
        .self$pass_to_next(drug)
      } else {
        message("Initial review already processed for ", drug$name)
        .self$pass_to_next(drug)
      }
    }
  )
)

# Concrete handler for Clinical Trials Review
ClinicalTrialsHandler <- setRefClass(
  "ClinicalTrialsHandler",
  contains = "DrugApprovalHandler",
  methods = list(
    handle_request = function(drug) {
      if (drug$clinical_trials_status == "Pending") {
        message("Clinical trials review passed for ", drug$name)
        drug$clinical_trials_status <- "Approved"
        .self$pass_to_next(drug)
      } else {
        message("Clinical trials already processed for ", drug$name)
        .self$pass_to_next(drug)
      }
    }
  )
)

# Concrete handler for Regulatory Review
RegulatoryReviewHandler <- setRefClass(
  "RegulatoryReviewHandler",
  contains = "DrugApprovalHandler",
  methods = list(
    handle_request = function(drug) {
      if (drug$regulatory_status == "Pending") {
        message("Regulatory review passed for ", drug$name)
        drug$regulatory_status <- "Approved"
        .self$pass_to_next(drug)
      } else {
        message("Regulatory review already processed for ", drug$name)
        .self$pass_to_next(drug)
      }
    }
  )
)

# Concrete handler for Final Approval
FinalApprovalHandler <- setRefClass(
  "FinalApprovalHandler",
  contains = "DrugApprovalHandler",
  methods = list(
    handle_request = function(drug) {
      if (drug$initial_review_status == "Approved" &&
          drug$clinical_trials_status == "Approved" &&
          drug$regulatory_status == "Approved") {
        message("Final approval granted for ", drug$name)
        drug$final_approval_status <- "Approved"
      } else {
        message("Drug ", drug$name, " could not pass final approval.")
      }
    }
  )
)

# Drug object
Drug <- setRefClass(
  "Drug",
  fields = list(
    name = "character",
    initial_review_status = "character",
    clinical_trials_status = "character",
    regulatory_status = "character",
    final_approval_status = "character"
  )
)

# Setup the chain of responsibility
initial_review <- InitialReviewHandler$new()
clinical_trials <- ClinicalTrialsHandler$new()
regulatory_review <- RegulatoryReviewHandler$new()
final_approval <- FinalApprovalHandler$new()

initial_review$set_next(clinical_trials)$set_next(regulatory_review)$set_next(final_approval)

drug1 <- Drug$new(
  name = "Test Small Molecule 01",
  initial_review_status = "Pending",
  clinical_trials_status = "Pending",
  regulatory_status = "Pending",
  final_approval_status = "Pending"
)

initial_review$handle_request(drug1)

File: Pharmaceutical_Company_Example/Behavioral/JavaScript/Chain_of_Responsibility.js
// Base class for a drug approval request handler
class DrugApprovalHandler {
  constructor() {
    this.nextHandler = null;
  }

  setNext(handler) {
    this.nextHandler = handler;
    return handler;
  }

  handleRequest(drug) {
    throw new Error("This method should be implemented by subclasses.");
  }

  passToNext(drug) {
    if (this.nextHandler) {
      this.nextHandler.handleRequest(drug);
    } else {
      console.log("End of chain, no handler could process the request.");
    }
  }
}

// Concrete handler for Initial Review
class InitialReviewHandler extends DrugApprovalHandler {
  handleRequest(drug) {
    if (drug.initial_review_status === "Pending") {
      console.log(`Initial review passed for ${drug.name}`);
      drug.initial_review_status = "Approved";
      this.passToNext(drug);
    } else {
      console.log(`Initial review already processed for ${drug.name}`);
      this.passToNext(drug);
    }
  }
}

// Concrete handler for Clinical Trials Review
class ClinicalTrialsHandler extends DrugApprovalHandler {
  handleRequest(drug) {
    if (drug.clinical_trials_status === "Pending") {
      console.log(`Clinical trials review passed for ${drug.name}`);
      drug.clinical_trials_status = "Approved";
      this.passToNext(drug);
    } else {
      console.log(`Clinical trials already processed for ${drug.name}`);
      this.passToNext(drug);
    }
  }
}

// Concrete handler for Regulatory Review
class RegulatoryReviewHandler extends DrugApprovalHandler {
  handleRequest(drug) {
    if (drug.regulatory_status === "Pending") {
      console.log(`Regulatory review passed for ${drug.name}`);
      drug.regulatory_status = "Approved";
      this.passToNext(drug);
    } else {
      console.log(`Regulatory review already processed for ${drug.name}`);
      this.passToNext(drug);
    }
  }
}

// Concrete handler for Final Approval
class FinalApprovalHandler extends DrugApprovalHandler {
  handleRequest(drug) {
    if (
      drug.initial_review_status === "Approved" &&
      drug.clinical_trials_status === "Approved" &&
      drug.regulatory_status === "Approved"
    ) {
      console.log(`Final approval granted for ${drug.name}`);
      drug.final_approval_status = "Approved";
    } else {
      console.log(`Drug ${drug.name} could not pass final approval.`);
    }
  }
}

let drug1 = {
  name: "Test Small Molecule 01",
  initial_review_status: "Pending",
  clinical_trials_status: "Pending",
  regulatory_status: "Pending",
  final_approval_status: "Pending",
};

// Setup the chain of responsibility
const initialReview = new InitialReviewHandler();
const clinicalTrials = new ClinicalTrialsHandler();
const regulatoryReview = new RegulatoryReviewHandler();
const finalApproval = new FinalApprovalHandler();

initialReview.setNext(clinicalTrials).setNext(regulatoryReview).setNext(finalApproval);

initialReview.handleRequest(drug1);
